### ISSUE TITLE
roachtest: tpcc-1000 roachtests have failing checks

### ISSUE THREAD

**Issue by lucy-zhang** [cockroachdb/cockroach#34025]
Some roachtests that involve running tpcc-1000 are failing the tpcc checks at the end, e.g.:

```
		I190114 15:41:47.835436 1 workload/tpcc/tpcc.go:290  check 3.3.2.1 took 162.495391ms
		Error: check failed: 3.3.2.1: 12 rows returned, expected zero
```
The tests affected by this:

- #33624 (schemachange/index/tpcc-1000)
- #33335 (schemachange/indexrollback/tpcc-1000)
- #33151 (scrub/index-only/tpcc-1000)
- #33149 (scrub/all-checks/tpcc-1000)
- https://github.com/cockroachdb/cockroach/issues/32813#issuecomment-453920621 (cdc/tpcc-1000) is potentially related

https://teamcity.cockroachdb.com/viewLog.html?buildId=1085451&buildTypeId=Cockroach_Nightlies_WorkloadNightly&tab=buildResultsDiv seems to be the first test failure of this kind. I reproduced this failure on `scrub/index-only/tpcc-1000` on a separate roachprod cluster, and ran SCRUB on the entire tpcc database, which did not turn up any anomalies.

It's not clear whether this is a schema change problem, a problem with the tests, or something else. Some next steps would be to determine whether this occurs when just running tpcc-1000 by itself, and look at what the anomalous rows are in the results of the check query.

**Comment by vivekmenezes** [comment]
It's running 
`SELECT count(*) FROM
 (SELECT max(no_o_id) - min(no_o_id) - count(*) AS nod FROM new_order
  GROUP BY no_w_id, no_d_id)
 WHERE nod != -1
`

expecting 0 to be returned. The schema changes do not run on the new_order table which makes it especially confounding.

**Comment by jordanlewis** [comment]
Yeah I would say that this probably isn't a schema change issue per se. When we've seen issues like this before, it's been either a problem with the test or a problem with transactionality in the database itself.

**Comment by petermattis** [comment]
Do we run the TPC-C checks after the restore/import is finished? I'm wondering if this is possibly a problem with restore/import.

**Comment by jordanlewis** [comment]
I believe we do run the checks after the restore/import, right @lucy-zhang? I think we run two sets - after the restore/import and after the test, and this problem was after the test.

**Comment by lucy-zhang** [comment]
Yes, the checks are run immediately after the restore/import, and they pass.

**Comment by vivekmenezes** [comment]
https://teamcity.cockroachdb.com/project.html?projectId=Cockroach_Nightlies&testNameId=-6687189696048107282&tab=testDetails

starting failing on Jan 11th on master

https://teamcity.cockroachdb.com/project.html?projectId=Cockroach_Nightlies&testNameId=-2631868135262591507&tab=testDetails

started failing on Jan 14th on master

https://teamcity.cockroachdb.com/project.html?projectId=Cockroach_Nightlies&testNameId=9189024510331884225&tab=testDetails

started failing on Jan 14th on master

So it does appear something changed around Jan 10-11th that is causing this problem

**Comment by vivekmenezes** [comment]
https://github.com/cockroachdb/cockroach/pull/33566 was a complex change that got merged before the failure.

**Comment by tbg** [comment]
A different check failed here:

https://github.com/cockroachdb/cockroach/issues/34293

**Comment by nvanbenschoten** [comment]
Yeah, these all look related. I'm working on isolating #33566 as the culprit now. Once that's confirmed, I'll dig into what's going wrong.

**Comment by tbg** [comment]
SGTM, let me know if I can be of any help.

**Comment by vivekmenezes** [comment]
@nvanbenschoten thanks for your help on this issue!

**Comment by nvanbenschoten** [comment]
I just got a reproduction on c5516efba7bd06fad5a8f26a751aa70f74f755c7. That doesn't tell us too much since we saw it before, but it's nice to have some degree of reproducibility, even if it takes a few hours.

I convinced myself that the `duplicate key value (o_w_id,o_d_id,o_id)=(979,7,3002) violates unique constraint "primary" (SQLSTATE 23505)` errors are different symptoms for the same problem demonstrated by the check failures. For instance, 
```
Error: check failed: 3.3.2.2: inequality at idx 435: order: 3010.000000, newOrder: 3010.000000, district-1: 3009.000000
```
is just another way of saying the the `order` table has an order id (`3010`) that does not line up with the `d_next_o_id` from the `district` table (also `3010`, but should at least be `3011`). This should not be possible given a properly functioning `new_order` transaction.

```
Error: check failed: 3.3.2.1: 12 rows returned, expected zero
```
is similar in that it demonstrates a partially applied `payment` txn.

These two check failures also fit a similar pattern. In both, the first write in the transaction seems to be missing. Interestingly, in both of these transactions, the writes are meant to be the most heavily contended. I suspect that we're seeing something like the following series of events:
- transaction A writes intent X for statements 1
- transaction B begins blocking on transaction A, intent X because its trying to write to the same row
- transaction A writes intents Y and Z for statements 2 and 3
- transaction B succesfully aborts transaction A and removes intent X
- transaction A's aborted record is removed (GC?)
- transaction A somehow "comes back to life" and re-writes its record either through a HeartbeatRequest or an EndTransaction request
- transaction A commits and resolves Y and Z. It fails to find X, so it throws [this warning](https://github.com/cockroachdb/cockroach/blob/6284b51757a69446447ea9d3cf8c6168dc716343/pkg/storage/engine/mvcc.go#L2166) (or one of the ones below it)

Unfortunately I don't actually see a warning like that in the logs, but perhaps the transaction didn't commit at epoch 0. I'm going to add some logging in and see if I can get a repro again.

**Comment by nvanbenschoten** [comment]
An interesting note is that these `scrub` and `schemachange` roachtests are dramatically underprovisioned, which leads to them having extremely long transactions. They run TPC-C 1000 on 3-node clusters with 4 cores each. This cluster config can usually only support around 450 warehouses before falling over.

**Comment by nvanbenschoten** [comment]
I'm still not to the bottom of this, but it's been my main focus for about a week and feel like I'm slowly painting a full picture of what's going on. The slow reproduction cycle (1-2 times a day) doesn't make quick iteration easy.

First off, I'll retract my previous statement that the `duplicate key value` / `3.3.2.2` check failure is the same issue as what we see here. I've been stressing tpcc (through the roachtests listed in this issue) for the past six days on c5516efba7bd06fad5a8f26a751aa70f74f755c7 and have yet to see a single instance of this (granted I've only seen about 5 `3.3.2.1` failures with 16 clusters continuously stressing this). @awoods187 seems to be able to reproduce these at-will on `master` and we saw the first instance about a week after we first saw the first `3.3.2.1` failure. So the conclusion I've been pushed towards is that they're not the same. If anyone wants to help, investigating that in parallel would be *very* helpful.

In terms of the easiest way to reproduce this issue, I've found that the `scrub` tests fail most frequently. I believe this is because they most severely underprovision a cluster for tpcc, leading to more transaction aborts. This is supported by the failure being slightly easier to reproduce when I dropped `TxnLivenessThreshold` down to `1 * base.DefaultHeartbeatInterval` instead of `2 * base.DefaultHeartbeatInterval`, which should cause more transaction aborts due to delayed txn heartbeats.

Another interesting point which @danhhz and I discussed in https://github.com/cockroachdb/cockroach/issues/32813#issuecomment-459091208 is that the failure usually occurs within a minute of starting load on a tpcc cluster. By dropping the load duration in the scrub tests down to 10 minutes I was able to shorten the iteration time without reducing the failure rate (noticeably?). More on why I think this is later.

Given that our prime suspect for this is 04189f59930815d7367c434b2ad617cc3cf1fcfc (although other than https://github.com/cockroachdb/cockroach/issues/33149#issuecomment-452606469, I've never seen this fail before c5516efba7bd06fad5a8f26a751aa70f74f755c7), I started by adding a lot of logging around transaction state transitions. Specifically, I logged about transaction pushes, transaction aborts, transaction record creation, transaction commits, etc. When I finally got a reproduction, I parsed all the logs, threw them into an advanced system specializing in ad-hoc query processing over large data sets (a CockroachDB cluster), and began looking for strange transaction timelines. I didn't find anything that stood out and soon realized that I had been logging so heavily for ~3 hours that most of the log files had rolled over and I was missing anything of interest. After fixing that and waiting another day for a repro, I finally had a full set of logs. Unfortunately, on its own, I still couldn't find anything that looked strange.

Since I still had the cluster up and running, I could verify that the [query from check `3.3.2.1`](https://github.com/cockroachdb/cockroach/blob/1b8689c0b4df102e1bf4e271913c4bb096ca8ffe/pkg/workload/tpcc/checks.go#L29) still reproduced. The check confirmed that the year-to-date sum on a `warehouse` row equaled the combined year-to-date sum of all of that warehouse's districts. By adjusting it slightly, it was possible to determine the warehouse that was inconsistent and the price difference that it was inconsistent by. I took this and found that in most cases I could find a corresponding row in the `history` table with the same amount! I then realized just how brilliant AS OF SYSTEM TIME queries are. I could perform the query with AS OF SYSTEM TIME specifiers to determine exactly when this `history` row was inserted. Mapping that back to my log of transactions I was able to find the exact transaction that performed this insert. Since the warehouse table was missing this amount, I figured the transaction's intent on the warehouse table had somehow been removed prematurely. This was essentially the theory in https://github.com/cockroachdb/cockroach/issues/34025#issuecomment-458540345.

I spent a while looking at neighboring transactions in the logs and how they interacted with our suspect transaction (let's call it Txn A). Everything checked out. No one thought they had aborted A (other than after it had committed and its txn record had been removed, but that's kosher). I went back through the reproduction cycle twice more just to get more logging on intent resolution, intent removal, and txn record removal. Still, nothing popped up. I then realized I could even query the `warehouse` table immediately after the transaction was supposed to be committed using AOST and see the update. I found that the `warehouse` table **had** been correctly updated by Txn A. I then confirmed that the `3.3.2.1` check passed **after** Txn A was supposed to have committed. At this point, I was thoroughly confused.

This is when I did what I should have done in the first place and ran the `3.3.2.1` check using AOST to determine exactly when it started failing. This ended up being about 10 seconds after Txn A committed (which remember, updated w_ytd by exactly the amount it was later missing). I looked in the `history` table again and sure enough, found a row inserted by a transaction (call it Txn B) at the same time. It turns out that Txn B updated the `warehouse` row by exactly `TxnB.amount - TxnA.amount`. In other words, Txn B had completed ignore Txn A's original update to the `warehouse` row. This was classic write skew of two read-modify-write operations!

I've since seen this exact behavior on three different clusters (every repro since I started looking for it). I've even seen it four times on the same cluster once, which corresponded to a  `4 rows returned, expected zero` failure. Here are what I see to be the commonalities between all of these cases:
1. Txn A always commits with a **lower** commit timestamp than Txn B's original timestamp
2. Everyone always commits at epoch 0 (i.e. no restarts)
3. Txn A always seems to hit a `RETRY_SERIALIZABLE` error, so I assume it's refreshing to avoid restarting (see fact #2)
4. Txn B usually but not always hits a `RETRY_SERIALIZABLE` error. In other words, it doesn't always refresh.
5. Txn A always seems to be surrounded by the following operations on its range: Add replica, Remove replica, and one or more splits. I highly suspect that this is related because it would explain why we see this so frequently at the beginning of runs.
6. From what I can tell, Txn A usually seems to commit on a different leaseholder for its `warehouse` range (where it's txn record is) than the leaseholder who existed when it stated. However, I can see cases where its heartbeat creates a txn record on the first and second leaseholder. Unfortunately, I've never been able to see the exact lease history of the range.
7. In the `4 rows returned, expected zero` case, the Txn A in all fours cases operated at essentially the exact same time, on the same range.

Since Txn B always has a higher orig timestamp than Txn A's commit timestamp (although they are active concurrently) it's baffling that Txn B misses Txn A's write. What should happen is either
1.  (if the read comes after the intent write) Txn B observes Txn A's write. If Txn A has already resolved its intent then it should absolutely respect it. If Txn A has not, then Txn B would start pushing Txn A. If Txn B is not able to abort Txn A then it should observe its update after the intent is resolved.
2. (if the read comes before the intent write) Txn B bumps the timestamp cache to TxnB.OrigTimestamp after its read, preventing Txn A from writing without bumping TxnA.Timestamp to at least TxnB.OrigTimestamp. This timestamp bump does not happen (see fact #1).

I haven't been able to think of any way that the second mechanism could break down in general or due to any of the changes in 04189f59930815d7367c434b2ad617cc3cf1fcfc. Even if TxnB incorrectly thought that TxnA was aborted (which I don't see evidence for), I don't see evidence that it tried to remove TxnA's intent. Without removing the intent, I don't think it's possible for it to have read the previously committed value. So I think a breakdown in the second mechanism is more likely, especially because this seems to be related to lease changes. But I also don't see how that could break down.

Random theories I explored but disproved:
1. the write skew is a result of a follower read gone wrong. This would explain why Txn B's read would either miss Txn A's intent/committed value or fail to prevent the intent by bumping the read timestamp cache. It turns out that `kv.closed_timestamp.follower_reads_enabled` defaults to false.
2. Txn B performed the read portion of its UPDATE statement in a different txn (with a lower original timestamp) and performed the write portion of its UPDATE statement as Txn B. This idea is similar to what happened in https://github.com/cockroachdb/cockroach/pull/33312. Since I started looking for this, I have always seen Txn B be a result of a previously aborted txn, but that txn always tried to commit unsuccessfully with 6 intents before being aborted. This means that it couldn't have been aborted in the middle of the SQL transaction's UPDATE statement.

Any insight or wild guesses here would be helpful. This is going to delay the next Alpha (as will the `duplicate key value` error I mentioned above) if we can't fix it by then, so there is some urgency here.

cc. @tbg @bdarnell @petermattis

**Comment by petermattis** [comment]
This is a fantastic write-up, though I'm afraid I don't have any good ideas to offer. One thing to keep in mind is that it is possible that https://github.com/cockroachdb/cockroach/commit/04189f59930815d7367c434b2ad617cc3cf1fcfc simply revealed a previously existing bug (perhaps one that was not exercisable prior to that change). 

> 7. In the 4 rows returned, expected zero case, the Txn A in all fours cases operated at essentially the exact same time, on the same range.

That's super curious. Something happened on that range that affected 4 different txns? I suppose this lends weight to the suspicion that an add replica, remove replica, or split, is related to the problem.

**Comment by tbg** [comment]
I'll sleuth some more code and read your report more thoroughly tomorrow, but I'll let [this script](https://gist.github.com/tbg/8b6eaf00745f3731d604f1fbbd05ab9b) run overnight on my worker in the hope that it reproduces. It basically merges-splits-scatters `warehouses` repeatedly and checks 3.3.2.1 after each iteration. I have it on c5516efba7bd06fad5a8f26a751aa70f74f755c7 currently.

**Comment by tbg** [comment]
Very likely unrelated, but my tpcc workload always dies with (right at the beginning, takes a few sec though):

$ ./bin/workload run tpcc --expensive-checks --warehouses 100 --ramp 30s --wait=false --tolerate-errors --duration=24h 'postgres://root@localhost:26257?sslmode=disable' --histograms tpcc.json
Error: preparing
		INSERT INTO "order" (o_id, o_d_id, o_w_id, o_c_id, o_entry_d, o_ol_cnt, o_all_local)
		VALUES ($1, $2, $3, $4, $5, $6, $7): ERROR: root: memory budget exceeded: 10240 bytes requested, 134215680 currently allocated, 134217728 bytes in budget (SQLSTATE 53200)
```

That's 134mb.

**Comment by bdarnell** [comment]
> (if the read comes after the intent write) Txn B observes Txn A's write. If Txn A has already resolved its intent then it should absolutely respect it. If Txn A has not, then Txn B would start pushing Txn A. If Txn B is not able to abort Txn A then it should observe its update after the intent is resolved.

This gives me flashbacks to the time-bound iterator bug. I know there aren't any TBIs in use here, though, and it doesn't look like we can point the finger to any other rocksdb changes/features.

> One thing to keep in mind is that it is possible that 04189f5 simply revealed a previously existing bug (perhaps one that was not exercisable prior to that change).

Yeah, just like pipelined writes in 2.1 exposed #30792 even though the faulty code had existed since 2.0. 

> This is going to delay the next Alpha (as will the duplicate key value error I mentioned above) if we can't fix it by then, so there is some urgency here.

Can we revert the changes to `{Begin,Heartbeat,End}Txn` instead of delaying?


I'm thinking about the way we use the txn id payload of the tscache, and how it's different between regular KV entries (where the txn id indicates a txn that is allowed to read at that timestamp) and txn records (where the txn id indicates whether we finalized our own txn or were aborted by someone else). In either case, the txn id is nilled out on a lease transfer (and, as you pointed out when this was in code review, it's been like that for a long time - consistent with the theory that this is something like #30792, an old bug that was recently exposed). 

8143b458 removed the consultsTSCache flag from BeginTransaction, replacing it with the CanCreateTxnRecord method. That method has explicit logic around nil txn ids, but maybe it's incorrect. Or maybe we were unknowingly relying on the behavior of applyTSCache for BeginTransaction (which doesn't handle txn ids in quite the same way) and this broke something.

**Comment by tbg** [comment]
Am I missing something or is it completely unclear how any bug related to the transaction record (or its timestamp cache entries) would even cause this? If I'm reading the above correctly, then either txn B just ignores an intent (but doesn't remove it) or B's read of `w_ytd` is not reflected by the timestamp cache (allowing the intent to be laid down under a prior read).

**Comment by tbg** [comment]
Familiarizing myself with the debugging a little bit. Here are my steps:

Show the mismatches. (check 3.3.2.1)

```
SELECT *, w_ytd-sum_d_ytd  FROM warehouse FULL OUTER JOIN 
(SELECT d_w_id, sum(d_ytd) as sum_d_ytd FROM district GROUP BY d_w_id)
ON (w_id = d_w_id)  WHERE w_ytd != sum_d_ytd;
```

There are four mismatches in this run. This would correspond to 3.3.2.1 complaining about four rows.

Find a payment txn (that would become our txnA) that used this missing amount corresponding to first warehouse `w_id=481`.

```
root@:26257/tpcc> select * from history where h_amount = 4700.94;
                 rowid                 | h_c_id | h_c_d_id | h_c_w_id | h_d_id | h_w_id |          h_date           | h_amount |    h_data
+--------------------------------------+--------+----------+----------+--------+--------+---------------------------+----------+---------------+
  2cb52a1d-4841-4f80-8941-fd0589ca11ea |    719 |       10 |      481 |     10 |    481 | 2019-02-01 13:53:23+00:00 |  4700.94 | 8    qVVgR5IS
(1 row)
```

Look at the payments to the same warehouse following that one:

```
root@:26257/tpcc> select * from history where h_w_id = 481 and h_date >= '2019-02-01 13:53:23+00:00' order by h_date asc;
                 rowid                 | h_c_id | h_c_d_id | h_c_w_id | h_d_id | h_w_id |          h_date           | h_amount |    h_data
+--------------------------------------+--------+----------+----------+--------+--------+---------------------------+----------+---------------+
  08f0ed80-ff9a-4c8e-abc0-5ef701df00b6 |   1726 |        1 |      481 |      1 |    481 | 2019-02-01 13:53:23+00:00 |   975.46 | 8    u94AA9
  2cb52a1d-4841-4f80-8941-fd0589ca11ea |    719 |       10 |      481 |     10 |    481 | 2019-02-01 13:53:23+00:00 |  4700.94 | 8    qVVgR5IS
  95f2f306-9ae6-475a-8cea-59957c4fbfeb |   2549 |       10 |      481 |     10 |    481 | 2019-02-01 13:53:47+00:00 |  1092.37 | 8    qVVgR5IS
  94524706-2a76-4d2e-a0fb-18a4977bce5a |   2785 |        7 |      481 |      7 |    481 | 2019-02-01 13:54:22+00:00 |  4332.45 | 8    8oKLWx6
  ab2421a1-245d-4f2a-886a-2f633359ea8a |    687 |        9 |      481 |      9 |    481 | 2019-02-01 13:54:50+00:00 |   851.88 | 8    dXsHwI
  2b8cc047-9c86-458f-b935-4e4c2bb4277b |    649 |       10 |      474 |      1 |    481 | 2019-02-01 13:56:39+00:00 |  2992.77 | 8    u94AA9
  762fd911-6262-4d89-ac7d-2df21b5010e0 |   1964 |        4 |      481 |      4 |    481 | 2019-02-01 13:56:47+00:00 |  3422.36 | 8    zJNAbvm3
  a7a3e25e-86f8-4bc3-b1b0-b4f7a9da36df |   1747 |        2 |      385 |      3 |    481 | 2019-02-01 14:01:16+00:00 |  2227.50 | 8    lKga79r
  72ea79cf-d2d9-4265-bf18-47e780fd9774 |   2812 |        9 |      481 |      9 |    481 | 2019-02-01 14:02:53+00:00 |  1178.06 | 8    dXsHwI
  bc05d4cd-d636-4017-af5f-aa7652d23671 |   1175 |        3 |      473 |      6 |    481 | 2019-02-01 14:03:51+00:00 |  2838.90 | 8    0NEQV0
  e548b585-d031-423c-823f-3ee21e11cb1f |   1877 |        5 |      519 |      4 |    481 | 2019-02-01 14:04:23+00:00 |  3118.12 | 8    zJNAbvm3
  4d0e5cc2-24ba-4d47-9b0e-8e8fb2a0fd54 |    679 |        6 |      481 |      6 |    481 | 2019-02-01 14:04:30+00:00 |  4574.18 | 8    0NEQV0
  236bd1a9-dede-4656-bb78-0d3cffad7d24 |   2315 |        2 |      481 |      2 |    481 | 2019-02-01 14:04:37+00:00 |   557.95 | 8    NQbLhE
  91c8eb79-3f51-4687-a58b-1a3937e52411 |   2119 |        5 |      481 |      5 |    481 | 2019-02-01 14:05:22+00:00 |  2919.42 | 8    55McjO13
  7628d07f-bada-4793-af48-97803680fc9b |   1751 |        7 |      481 |      7 |    481 | 2019-02-01 14:05:53+00:00 |  3479.89 | 8    8oKLWx6
  6a95ac00-c62c-4279-9691-bab1be6b39ac |   1861 |        4 |      481 |      4 |    481 | 2019-02-01 14:07:12+00:00 |  2771.83 | 8    zJNAbvm3
  726cecd0-2003-43ca-8921-c248ab4bf1c2 |    956 |        8 |      481 |      8 |    481 | 2019-02-01 14:07:14+00:00 |  4070.23 | 8    4Nc6ISb
  1573217f-a81e-47e5-a22d-97130b69ede1 |    699 |        7 |      481 |      7 |    481 | 2019-02-01 14:07:30+00:00 |  2275.37 | 8    8oKLWx6
  db52eeef-9f80-436f-8a03-e244901fce93 |    974 |        3 |      481 |      3 |    481 | 2019-02-01 14:07:31+00:00 |  4883.84 | 8    lKga79r
  f78d0d14-349a-4ea7-b7b8-e91a1ed961a0 |    655 |        3 |      481 |      3 |    481 | 2019-02-01 14:07:41+00:00 |  3744.82 | 8    lKga79r
  c5fa99aa-b94d-459c-8766-25e76d92e0f1 |   1827 |        5 |      481 |      5 |    481 | 2019-02-01 14:08:18+00:00 |  4045.97 | 8    55McjO13
(21 rows)
```

I find it curious that exactly the row that later goes unaccounted for at the warehouse level is double-booked (i.e. another row has the same timestamp), but maybe I just don't know how things work.

Anyway, the sum of all payments to that warehouse in the history should sum up to the sum of w_ytds of the district tables (it does)

```
root@:26257/tpcc> select sum(h_amount) from history where h_w_id = 481;
     sum
+-----------+
  390476.21
(1 row)
```

Time travelling back to the row's timestamp, the warehouse aggregate is actually correct:

```
root@:26257/tpcc> select w_ytd from warehouse as of system time '2019-02-01 13:53:23+00:00'                                                                                                                                                    where w_id = 481;
    w_ytd
+-----------+
  312319.77
(1 row)

Time: 885.962µs                                                                                                                                                                                      root@:26257/tpcc> select sum(h_amount) from history as of system time '2019-02-01 13:53:23+00:00' where h_w_id = 481;
     sum
+-----------+
  312319.77
(1 row)

Time: 661.285759ms
```

They still agree 10s after that timestamp, but not 20s:

```
root@:26257/tpcc> select sum(h_amount) from history as of system time '2019-02-01 13:53:53+00:00' where h_w_id = 481;
     sum
+-----------+
  320970.56
(1 row)

Time: 711.332639ms

root@:26257/tpcc> select w_ytd from warehouse as of system time '2019-02-01 13:53:53+00:00' where w_id = 481;
    w_ytd
+-----------+
  316269.62
(1 row)

root@:26257/tpcc> select 320970.56 - 316269.62;
  ?column?
+----------+
   4700.94
(1 row)
```

and we can look up in the table above that the only payment in that window of time was at `2019-02-01 13:53:47+00:00`, so this becomes txnB. At that point, I suppose, we need to dig in the logs.

**Comment by nvanbenschoten** [comment]
> Can we revert the changes to {Begin,Heartbeat,End}Txn instead of delaying?

Yes, I'll get a PR out. I'm pretty confident at this point that the worst that change does is make this more easy to hit, but for qualifying the Alpha maybe that's enough.

I did get a new reproduction and am seeing ... something. As always, I see the usual situation with TxnA and TxnB. The only difference from my previous observation is that neither ever hit a `RETRY_SERIALIZABLE` error, so we can ignore issues with txn refreshes.

Most of the new logging I added was around txn reads and lease change. Here's where things get interesting. First off, I see both TxnA and TxnB perform their initial read (where the skew happens) on leaseholder X. Later on, on leaseholder Y we see Txn A abort a separate TxnC. This must be before TxnA writes its first intent (on the `warehouse` table) because that's the only place where the TxnA and TxnC overlap. This means that TxnB reads before TxnA writes. So we're seeing some kind of failure with the timestamp cache and leaseholder changes (see `if the read comes before the intent write` above).

The next interesting point is that the leaseholder Y's replica was added to the range immediately before this. It then receives a snapshot shortly after TxnA and TxnB read. I can tell that this replica becomes a leaseholder only ~200us after this because I see leaseholder Y accepting requests that require it to be the leaseholder immediately after the snapshot:
```
nathan-1549267487-scrub-all-checks-tpcc-1000-0004> I190204 08:40:45.120271 401945 storage/replica_raftstorage.go:815  [n3,s3,r4767/4:/Table/59/1/{553/0-664}] applied Raft snapshot in 13ms [clear=0ms batch=0ms entries=1ms commit=10ms]
nathan-1549267487-scrub-all-checks-tpcc-1000-0004> E190204 08:40:45.120431 402412 storage/batcheval/cmd_heartbeat_txn.go:80  [n3,s3,r4767/4:/Table/59/1/{553/0-664}] NATHAN heartbeat create succeeded "sql txn" id=bcbc513c key=/Table/59/1/621/0 rw=true pri=0.06655999 stat=PENDING epo=0 ts=1549269609.342955168,0 orig=1549269609.342955168,0 max=1549269609.842955168,0 wto=false seq=0
```

However, I can't actually tell when the lease begins because my logging that fires when `leasePostApply.iAmTheLeaseHolder` is true fails to fire! In fact, the first time I see this fire is during a lease extension ~20 seconds later.
```
nathan-1549267487-scrub-all-checks-tpcc-1000-0004> I190204 08:41:03.617368 117 storage/replica_proposal.go:213  [n3,s3,r4767/4:/Table/59/1/{553/0-664}] NATHAN new range lease repl=(n3,s3):3 seq=4 start=1549269641.515388613,0 epo=1 pro=1549269641.515394166,0 following repl=(n3,s3):3 seq=4 start=1549269641.515388613,0 epo=1 pro=1549269641.515394166,0
```

We can see a few things from this log:
1. the log is for a lease extension
2. the lease started at `2019-02-04 08:40:41.515388613` and was proposed at `2019-02-04 08:40:41.515394166`
3. this start timestamp looks to be before the timestamp of TxnA and TxnB's reads (`190204 08:40:41.806886` and `190204 08:40:41.589374`), but I'm not sure that I trust that this is true, because one source is pulling from the log timestamp and one is pulling from and hlc timestamp.

So, I see three possibilities for what's going on here:
1. leaseholder X is allowing a read after it has started its lease transfer. Comparing the log timestamp to the hlc timestamp on the lease would indicate this, but I'm not convinced about this because I don't see any other evidence for it, so I suspect that the timestamp difference is a red herring.
2. `leasePostApply` is not being called on leaseholder Y because it receives a snapshot which includes it as the leaseholder. Since [this](https://github.com/cockroachdb/cockroach/blob/317226b32df6abc63987c27b4bd8eb75aaca94f5/pkg/storage/replica_proposal.go#L241) is the only place that we bump the timestamp cache, this would allow for a timestamp cache regression which could cause this issue.
3. `leasePostApply` is being called but [`iAmTheLeaseHolder`](https://github.com/cockroachdb/cockroach/blob/317226b32df6abc63987c27b4bd8eb75aaca94f5/pkg/storage/replica_proposal.go#L201) is not being detected, possibly because the new replica does not yet know its replica ID. Again, this could cause a timestamp cache regression.

The logs indicate that either 2 or 3 is the issue. I suspect 2 because I was just able to trigger an assertion that a snapshot never includes the recipient as the new leaseholder. I'm verifying now that the assertion fires even when I include the condition that the lease must _change_ due to the snapshot. I think this will be easy to repro because the assertion will fire even when it wouldn't have caused an anomaly.

**Comment by nvanbenschoten** [comment]
> 1. leaseholder X is allowing a read after it has started its lease transfer. Comparing the log timestamp to the hlc timestamp on the lease would indicate this, but I'm not convinced about this because I don't see any other evidence for it, so I suspect that the timestamp difference is a red herring.

Also, given the synchronization in `Replica`, this isn't necessarily a problem. It would only be a problem if the read's timestamp was not included in the local clock before the least transfer started. We can't tell whether that's true from the logs. In addition, I'm logging _after_ the read completes.

**Comment by nvanbenschoten** [comment]
Here's some more evidence that this a breakdown in the timestamp cache update in new leaseholders. From the logs above, we see that the first operation on the new leaseholder Y is to create transaction record due to a heartbeat:
```
nathan-1549267487-scrub-all-checks-tpcc-1000-0004> E190204 08:40:45.120431 402412 storage/batcheval/cmd_heartbeat_txn.go:80  [n3,s3,r4767/4:/Table/59/1/{553/0-664}] NATHAN heartbeat create succeeded "sql txn" id=bcbc513c key=/Table/59/1/621/0 rw=true pri=0.06655999 stat=PENDING epo=0 ts=1549269609.342955168,0 orig=1549269609.342955168,0 max=1549269609.842955168,0 wto=false seq=0
```
Notice the orig timestamp of this txn - `1549269609.342955168,0`.

We also see that the lease for replica Y started at - `1549269641.515388613,0`. This is above the transaction record's orig timestamp. This shouldn't be allowed. The timestamp cache on replica Y should have been bumped to `1549269641.515388613,0`, which should prevent a new transaction record with an orig timestamp beneath this time.

This is the first situation I see where https://github.com/cockroachdb/cockroach/commit/c5516efba7bd06fad5a8f26a751aa70f74f755c7 could be tickling this bug more easily. For instance, if a transaction is pushed or aborted before its record is first written, we now rely on the timestamp cache to convey this info. If lease transfers aren't preserving monotonicity in the timestamp cache, we could see lost txn pushes and aborts.

**Comment by nvanbenschoten** [comment]
Just saw the assertion fire again:
```
F190204 19:47:56.165916 928512 storage/replica_raftstorage.go:975  [n3,s3,r5422/4:/Table/54/1/9{41-75/0}] NATHAN changing lease to me in snapshot new=repl=(n3,s3):4 seq=5 start=1549309674.414592041,0 epo=1 pro=1549309674.414596472,0 old=repl=(n1,s1):2 seq=4 start=1549309552.633202959,0 epo=1 pro=1549309552.633219470,0
```

Notice that the lease is changing hands due to the snapshot. So it looks like `leasePostApply` is being supressed by snapshot (possibility 2 from above). Is the fix as straightforward as calling `leasePostApply` in `applySnapshot` (with a fake previous lease)?

**Comment by nvanbenschoten** [comment]
I wonder if forgetting to go through the `leasePostApply` steps after a snapshot materializes as other issues as well. For instance, we don't call `maybeWatchForMerge` in cases when we probably should. We also don't enable our `txnWaitQueue` or attempt to acquire raft leadership. All bad stuff that could cause surprising behavior.

**Comment by tbg** [comment]
> Is the fix as straightforward as calling leasePostApply in applySnapshot (with a fake previous lease)?

Something like that sounds right.

This is quite the uber bug, great work tracking that down. Just to make sure, clock offset is completely off the table now, right? We're seeing a pretty empty timestamp cache on the new leaseholder so it returns pretty old timestamps unless it rotates through the pages very quickly.

> This is the first situation I see where c5516ef could be tickling this bug more easily. For instance, if a transaction is pushed or aborted before its record is first written, we now rely on the timestamp cache to convey this info. If lease transfers aren't preserving monotonicity in the timestamp cache, we could see lost txn pushes and aborts.

I still struggle to connect the dots. If a snapshot-lease-transfer-included occurs, the window for this bug is open, right? You're suggesting that there could be all sorts of anomalies since we're relying on the ts cache more, and that's right, but we haven't actually seen them in any of your repros, or do you suspect that we have?

Have you ever tried to repro with the parent of your main testing commit? Maybe we'll find that it wasn't that commit after all.

**Comment by bdarnell** [comment]
Yes, nice detective work. This does seem likely to be the problem based on the evidence here. 

The biggest remaining question I have is why we're transferring a lease to a node that's behind in the first place. Even with the leasePostApply bug fixed, that's a performance problem. Is filterBehindReplicas failing somehow? Is Allocator.TransferLeaseTarget perhaps doing the wrong thing if filterBehindReplicas returns an empty list? (which it does in leader-not-leaseholder situations). 

Maybe the issue is exacerbated by #30716, which adds the range to the replicate queue when a new node becomes raft leader. This new leader has incomplete information and will immediately consider transferring the lease away. #31894 also caught my eye as a recent change in this area.

**Comment by nvanbenschoten** [comment]
The (likely) patch for this is ready for review - https://github.com/cockroachdb/cockroach/pull/34548.

> The biggest remaining question I have is why we're transferring a lease to a node that's behind in the first place. Even with the leasePostApply bug fixed, that's a performance problem. Is filterBehindReplicas failing somehow? Is Allocator.TransferLeaseTarget perhaps doing the wrong thing if filterBehindReplicas returns an empty list? (which it does in leader-not-leaseholder situations).

These clusters are so [overloaded](https://github.com/cockroachdb/cockroach/issues/34025#issuecomment-458574374) that it's believable that the Replica isn't behind when the lease transfer is proposed but falls behind before it is applied.

**Comment by tbg** [comment]
> These clusters are so [overloaded](https://github.com/cockroachdb/cockroach/issues/34025#issuecomment-458574374) that it's believable that the Replica isn't behind when the lease transfer is proposed but falls behind before it is applied.

That still seems pretty unlikely, but seems what's happening. We do call this in both paths

https://github.com/cockroachdb/cockroach/blob/4aa0b528dcef2c6a729abc32efaf7a5ce17a074d/pkg/storage/allocator.go#L1170

and that method leans towards false negatives.

**Comment by petermattis** [comment]
I just audited the filtering of behind replicas and lease transfer and everything looks kosher. The one thing I'm wondering is the special case for declaring the Raft leader is not behind. That shouldn't be a problem as we require the leaseholder and Raft leader to be co-located before performing lease transfers.

**Comment by tbg** [comment]
The only way I can picture this happening is that all the checks are passed, then we propose the command, get on some overlapping command (?) while other commands get proposed and so the log position that the command ends up in is more advanced than our view of the raft status.

Actually the Raft status might be pretty stale if the machine is really aching. It only changes when a Ready is proposed. If the disks are slow etc there might be a large ready that commits a bunch of stuff that the soon-to-be-leaseholder wasn't involved in committing.

**Comment by bdarnell** [comment]
> These clusters are so [overloaded](https://github.com/cockroachdb/cockroach/issues/34025#issuecomment-458574374) that it's believable that the Replica isn't behind when the lease transfer is proposed but falls behind before it is applied.

But once the lease transfer is proposed, we stop proposing other writes. It shouldn't be able to fall further behind while the lease transfer is in flight.

**Comment by tbg** [comment]
@bdarnell my comment would explain this, but I can't say that I'll be fully satisfied with the explanation until I see it in practice.  Using Nathan's assertion with logging around lease transfers should do the trick.

**Comment by tbg** [comment]
Hooray!

**Comment by nvanbenschoten** [comment]
For completeness, I spent about 6 hours stressing the test with c5516efba7bd06fad5a8f26a751aa70f74f755c7 + the patch from https://github.com/cockroachdb/cockroach/pull/34548. I didn't see a single failure.

I then did the same test on `master`. Here I stumbled upon some good news and some bad news. The bad news is that I saw the check failure on one of my clusters. The good news is that I saw the same failure on every single other cluster! The test actually reliably fails on `master`.

The test still took 2 hours to run, but now that its failures were frequent, I was able to bisect the problem to 80eea9e35a541d5603111c436d5639125bd0c104. So it appears that we messed something up in that commit. The next step here is to figure out what that is.

This is actually in some senses a huge relief because it explains why the frequency of these failures (along with a few others with similar symptoms) has appeared to rise over the month of January. Amusingly, I mentioned something along these lines in https://github.com/cockroachdb/cockroach/issues/34025#issuecomment-460016784 - check out the dates between https://github.com/cockroachdb/cockroach/pull/33381#issuecomment-453863859 and https://github.com/cockroachdb/cockroach/pull/33396#issuecomment-452115855 :smile:

So it looks like the test failed very rarely before 80eea9e35a541d5603111c436d5639125bd0c104 because of the bug that was fixed today, but began failing regularly because of something in that commit. I feel a little silly that I never once thought to test on master before looking back at commits in the vacinity of where we saw the test first fail. On the bright side, it seems like this will lead to two bug fixes instead of just one, including one that has been lingering for years! I think that warrants another shout out to @lucy-zhang for introducing the test.

Of course, I think this also means we'll need to add whatever fix we create into the alpha that we cut today. Or we can just revert 80eea9e35a541d5603111c436d5639125bd0c104, as it was just a refactor and I don't think there's very much built on top of it.

cc. @andreimatei

**Comment by nvanbenschoten** [comment]
So with all that said, don't be surprised if we still see test failures tonight.

**Comment by tbg** [comment]
I wasn't able to find anything wrong in 80eea9e (though I should look again), but I figured that maybe we're returning ambiguous results more than before and that messes something up in higher layers. Not really sure this is related to this particular bug, but it seems weird that the DistSender code below returns ambiguous results either as `br.Err` or as `err`.

https://github.com/cockroachdb/cockroach/blob/70be833b48d79a7ead1608f30d30e4b38765e8e4/pkg/kv/dist_sender.go#L1475-L1484

**Comment by tbg** [comment]
Hm, seems like both cases get joined later:

https://github.com/cockroachdb/cockroach/blob/70be833b48d79a7ead1608f30d30e4b38765e8e4/pkg/kv/dist_sender.go#L496-L512

**Comment by tbg** [comment]
Heh, this is more likely the cause of the bug:

![image](https://user-images.githubusercontent.com/5076964/52328072-a47b2580-29ee-11e9-9812-3694ceb7eb2e.png)

note how `response.Err` is used to determine whether to hand out the intents to the caller. I think what that could do is reinstantiate a bug we'v fixed before: if an EndTransaction evaluates successfully but bounces below Raft because of an illegal lease index, we will now give the committed proto to waiting pushers (or similar during a PushTxn that aborts). I'll see if that fixes it on master.

**Comment by tbg** [comment]
Running two instances of scrub-index-only-tpcc-1000 on `master` and two on a binary that has the one-line fix (replacing `response.Err` by `pErr` in the highlighted line above).

**Comment by tbg** [comment]
The first failure on `master` is in:

```
--- FAIL: scrub/index-only/tpcc-1000 (10829.34s)
	test.go:743,cluster.go:1226,tpcc.go:130,scrub.go:58: /Users/tschottdorf/go/bin/roachprod run tobias-1549443202-scrub-index-only-tpcc-1000:5 -- ./workload check tpcc --warehouses=1000 {pgurl:1} returned:
		stderr:
		I190206 11:53:20.623699 1 workload/tpcc/tpcc.go:288  check 3.3.2.1 took 2.090668716s
		Error: check failed: 3.3.2.1: 9 rows returned, expected zero
		Error:  exit status 1

		stdout:
		: exit status 1
```

the second `master` instance is still running. The two fixed instances have already made it past check 3.3.2.1 🙌 and probably even all the checks (the test is "waiting" which I assume is the scrub part).

So assuming the second master run also fails, I'm fairly comfortable that that was the bug.

**Comment by tbg** [comment]
and here's the second failure on `master`:

```
--- FAIL: scrub/index-only/tpcc-1000 (11138.28s)
	test.go:743,cluster.go:1226,tpcc.go:130,scrub.go:58: /Users/tschottdorf/go/bin/roachprod run tobias-1549443223-scrub-index-only-tpcc-1000:5 -- ./workload check tpcc --warehouses=1000 {pgurl:1} returned:
		stderr:
		I190206 11:58:48.216353 1 workload/tpcc/tpcc.go:288  check 3.3.2.1 took 3.921505659s
		Error: check failed: 3.3.2.1: 7 rows returned, expected zero
		Error:  exit status 1

		stdout:
		: exit status 1
```

**Comment by bdarnell** [comment]
Fool me twice, shame on me. This definitely reveals a flaw in our testing. We added a test #32236 for #30792, but apparently it was too narrowly-targeted and wasn't sensitive to other related bugs. We should think about a more general way to test for things like this. I'd have thought the jepsen bank test would at least be sensitive to this kind of thing.

One thought is to run tests with a flag that turns any reevaluations into non-retryable errors. This means that any "leakage" like this would result in long-lasting inconsistencies, instead of only a very short-lived window in which the side effects are applied before the command ultimately succeeds after retry.
